cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/DefaultPreviewRunnerManager.java

```diff
@@ -0,0 +1,216 @@
+    // Create and start the preview poller services.
+    for (int i = 0; i < maxConcurrentPreviews; i++) {
+      String pollerInfo = UUID.randomUUID().toString();
+
+      PreviewRunnerService pollerService = new PreviewRunnerService(
+        previewCConf, previewInjector.getInstance(PreviewRunner.class),
+        previewRequestFetcherFactory.create(Bytes.toBytes(pollerInfo)));
+
+      pollerService.startAndWait();
+      previewPollers.put(pollerInfo, pollerService);
+    }
+    PreviewRunner runner = previewInjector.getInstance(PreviewRunner.class);
+    if (runner instanceof Service) {
+      ((Service) runner).startAndWait();
+    }
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    PreviewRunner runner = previewInjector.getInstance(PreviewRunner.class);
+    if (runner instanceof Service) {
+      stopQuietly((Service) runner);
+    }
+
+    for (Service pollerService : previewPollers.values()) {
+      stopQuietly(pollerService);
+    }
+  }
+
+  private void stopQuietly(Service service) {
+    try {
+      service.stopAndWait();
+    } catch (Exception e) {
+      LOG.warn("Error stopping the preview runner.", e);
+    }
+  }
+
+  @Override
+  public void stop(byte[] runnerId) throws Exception {
+    PreviewRunnerService service = previewPollers.get(Bytes.toString(runnerId));
+    if (service == null) {
+      throw new NotFoundException("Preview run cannot be stopped. Please try stopping again or start new preview run.");
+    }
+    service.stopAndWait();
+    String newRunnerId = UUID.randomUUID().toString();
+    PreviewRunnerService newService = new PreviewRunnerService(
+      previewCConf, previewInjector.getInstance(PreviewRunner.class),
+      previewRequestFetcherFactory.create(Bytes.toBytes(newRunnerId)));
+    newService.startAndWait();
+    previewPollers.put(newRunnerId, newService);
+  }
+
+  /**
+   * Create injector for the given application id.
+   */
+  @VisibleForTesting
+  Injector createPreviewInjector() throws IOException {
+    return Guice.createInjector(
+      new ConfigModule(previewCConf, previewHConf, previewSConf),
+      new IOModule(),
+      new AuthenticationContextModules().getMasterModule(),
+      new PreviewSecureStoreModule(secureStore),
+      new PreviewDiscoveryRuntimeModule(discoveryService),
+      new LocalLocationModule(),
+      new ConfigStoreModule(),
+      previewRunnerModule,
+      new ProgramRunnerRuntimeModule().getStandaloneModules(),
+      new PreviewDataModules().getDataFabricModule(transactionSystemClient, previewLevelDBTableService),
+      new PreviewDataModules().getDataSetsModule(datasetFramework),
+      new DataSetServiceModules().getStandaloneModules(),
+      // Use the in-memory module for metrics collection, which metrics still get persisted to dataset, but
+      // save threads for reading metrics from TMS, as there won't be metrics in TMS.
+      new MetricsClientRuntimeModule().getInMemoryModules(),
```

Comment:
So where does the metrics go?? Inside the preview runner pod?

Reply:
Currently yes. Will add this change in separate PR.